server/src/test/java/org/elasticsearch/action/support/CancellableFanOutTests.java

```diff
@@ -130,4 +142,102 @@ protected String onCompletion() {
     }
+
+    public void testConcurrency() throws InterruptedException {
+
+        final var isProcessorThread = startsWith("processor-thread-");
+        final var isCancelThread = equalTo("cancel-thread");
+
+        final var items = randomList(1, 3, () -> randomAlphaOfLength(5));
+        final var processorThreads = new Thread[items.size()];
+        final var queue = new LinkedBlockingQueue<Runnable>();
+        final var barrier = new CyclicBarrier(processorThreads.length + 1);
+        for (int i = 0; i < processorThreads.length; i++) {
+            processorThreads[i] = new Thread(() -> {
+                try {
+                    assertThat(Thread.currentThread().getName(), isProcessorThread);
+                    final var item = Objects.requireNonNull(queue.poll(10, TimeUnit.SECONDS));
+                    safeAwait(barrier);
+                    item.run();
+                } catch (Exception e) {
+                    throw new AssertionError(e);
+                }
+            }, "processor-thread-" + i);
+            processorThreads[i].start();
+        }
+
+        final var task = new CancellableTask(1, "test", "test", "", TaskId.EMPTY_TASK_ID, Map.of());
+
+        final var cancelThread = new Thread(() -> {
+            assertThat(Thread.currentThread().getName(), isCancelThread);
+            safeAwait(barrier);
+            TaskCancelHelper.cancel(task, "test");
+        }, "cancel-thread");
+        cancelThread.start();
+
+        final var itemsProcessed = new AtomicInteger();
+        final var completionLatch = new CountDownLatch(1);
+        new CancellableFanOut<String, String, String>() {
+            @Override
+            protected void sendItemRequest(String s, ActionListener<String> listener) {
+                queue.add(
+                    randomBoolean() ? () -> listener.onResponse(s) : () -> listener.onFailure(new ElasticsearchException("sendItemRequest"))
+                );
+            }
+
+            @Override
+            protected void onItemResponse(String s, String response) {
+                assertThat(Thread.currentThread().getName(), isProcessorThread);
+                assertEquals(s, response);
+                assertThat(itemsProcessed.incrementAndGet(), lessThanOrEqualTo(items.size()));
+            }
+
+            @Override
+            protected void onItemFailure(String s, Exception e) {
+                assertThat(Thread.currentThread().getName(), isProcessorThread);
+                assertThat(e, instanceOf(ElasticsearchException.class));
+                assertEquals("sendItemRequest", e.getMessage());
+                assertThat(itemsProcessed.incrementAndGet(), lessThanOrEqualTo(items.size()));
+            }
+
+            @Override
+            protected String onCompletion() {
+                assertEquals(items.size(), itemsProcessed.get());
+                assertThat(Thread.currentThread().getName(), anyOf(isProcessorThread));
```

Comment:
I think we can do without `anyOf` here?

```suggestion
                assertThat(Thread.currentThread().getName(), isProcessorThread);
```